edb/schema/objects.py

```diff
@@ -43,52 +47,113 @@
 from . import _types
 
 
-def get_known_type_id(typename, default=...):
+if TYPE_CHECKING:
+    from edb.schema import objtypes
+    from edb.schema import delta as sd
+    from edb.schema import schema as s_schema
+
+    if sys.version_info <= (3, 7):
+        from typing_extensions import Protocol  # type: ignore
+
+    CovT = TypeVar("CovT", covariant=True)
+
+    class MergeFunction(Protocol):
+        def __call__(
+            self,  # not actually part of the signature
+            target: InheritingObjectBase,
+            sources: List[Object],
+            field_name: str,
+            *,
+            schema: s_schema.Schema,
+        ) -> Any:
+            ...
+
+    class CollectionFactory(Collection[CovT], Protocol):
+        """An unknown collection that can be instantiated from an iterable."""
+
+        def __init__(self, from_iter: Optional[Iterable[CovT]] = None) -> None:
+            ...
+
+
+class NoDefaultT(enum.Enum):
+    """Used as a sentinel indicating that a named argument wasn't passed.
+
+    Trick from https://github.com/python/mypy/issues/7642.
+    """
+    NoDefault = 0
+
+
+NoDefault: Final = NoDefaultT.NoDefault
+T = TypeVar("T")
```

Comment:
Should lines 87..93 be moved up into the `if TYPE_CHECKING:` block?

Reply:
Some of them are used in the runtime context, e.g. in casts, type aliases and in generics that are not inside annotations. But I see that not all, I'll move 'em.